Mathlib/Logic/Function/Basic.lean

```diff
@@ -280,3 +281,401 @@ theorem partial_inv_left {α β} {f : α → β} (I : injective f) : ∀ x, part
+dif_neg h
+
+lemma forall_update_iff (f : ∀a, β a) {a : α} {b : β a} (p : ∀a, β a → Prop) :
+  (∀ x, p x (update f a b x)) ↔ p a b ∧ ∀ x, x ≠ a → p x (f x) :=
+Iff.intro
+  (by intro h
+      have h1 := h a
+      have h2 : update f a b a = b := update_same _ _ _
+      rw [h2] at h1
+      refine ⟨h1, ?_⟩
+      intro x hx
+      have h3 := update_noteq hx b f
+      rw [←h3]
+      exact h x)
+  (by intro ⟨hp,h⟩ x
+      have h1 : x = a ∨ x ≠ a := Decidable.em _
+      match h1 with
+      | Or.inl he => rw [he, update_same]
+                     exact hp
+      | Or.inr hne => have h4 := update_noteq hne b f
+                      rw [h4]
+                      exact h x hne)
+
+lemma update_eq_iff {a : α} {b : β a} {f g : ∀ a, β a} :
+  update f a b = g ↔ b = g a ∧ ∀ x, x ≠ a -> f x = g x :=
+funext_iff.trans $ forall_update_iff _ (λ x y => y = g x)
+
+lemma eq_update_iff {a : α} {b : β a} {f g : ∀ a, β a} :
+  g = update f a b ↔ g a = b ∧ ∀ x, x ≠ a -> g x = f x :=
+funext_iff.trans $ forall_update_iff _ (λ x y => g x = y)
+
+@[simp] lemma update_eq_self (a : α) (f : ∀a, β a) : update f a (f a) = f :=
+update_eq_iff.2 ⟨rfl, λ _ _ => rfl⟩
+
+lemma update_comp_eq_of_forall_ne' {α'} (g : ∀ a, β a) {f : α' → α} {i : α} (a : β i)
+  (h : ∀ x, f x ≠ i) :
+  (λ j => (update g i a) (f j)) = (λ j => g (f j)) :=
+funext $ λ x => update_noteq (h _) _ _
+
+/-- Non-dependent version of `function.update_comp_eq_of_forall_ne'` -/
+lemma update_comp_eq_of_forall_ne {α β : Sort _} (g : α' → β) {f : α → α'} {i : α'} (a : β)
+  (h : ∀ x, f x ≠ i) :
+  (update g i a) ∘ f = g ∘ f :=
+update_comp_eq_of_forall_ne' g a h
+
+lemma update_comp_eq_of_injective' (g : ∀a, β a) {f : α' → α} (hf : Function.injective f)
+  (i : α') (a : β (f i)) :
+  (λ j => update g (f i) a (f j)) = update (λ i => g (f i)) i a :=
+eq_update_iff.2 ⟨update_same _ _ _, λ j hj => update_noteq (hf.ne hj) _ _⟩
+
+/-- Non-dependent version of `function.update_comp_eq_of_injective'` -/
+lemma update_comp_eq_of_injective {β : Sort _} (g : α' → β) {f : α → α'}
+  (hf : Function.injective f) (i : α) (a : β) :
+  (Function.update g (f i) a) ∘ f = Function.update (g ∘ f) i a :=
+update_comp_eq_of_injective' g hf i a
+
+lemma apply_update {ι : Sort _} [DecidableEq ι] {α β : ι → Sort _}
+  (f : ∀i, α i → β i) (g : ∀i, α i) (i : ι) (v : α i) (j : ι) :
+  f j (update g i v j) = update (λ k => f k (g k)) i (f i v) j :=
+by by_cases h : j = i
+   subst j; simp
+   simp[h]
+
+lemma comp_update {α' : Sort _} {β : Sort _} (f : α' → β) (g : α → α') (i : α) (v : α') :
+  f ∘ (update g i v) = update (f ∘ g) i (f v) :=
+funext $ apply_update _ _ _ _
+
+theorem update_comm {α} [DecidableEq α] {β : α → Sort _}
+  {a b : α} (h : a ≠ b) (v : β a) (w : β b) (f : ∀a, β a) :
+  update (update f a v) b w = update (update f b w) a v :=
+by funext c
+   simp only [update]
+   by_cases h₁ : c = b <;> by_cases h₂ : c = a
+
+   -- case 1
+   rw [dif_pos h₁, dif_pos h₂]
+   cases h (h₂.symm.trans h₁)
```

Comment:
I believe we now have a fancy bullet points to structure tactic proofs, I hope this is the correct syntax:
```suggestion
   - rw [dif_pos h₁, dif_pos h₂]
     cases h (h₂.symm.trans h₁)
```

Reply:
Cool! I needed to add some parentheses to make it work: 85838b2fda9bc17803df0a15b4abc21e7cc24e1e

Reply:
This comment is what clued me in to the need for parens: https://github.com/leanprover-community/mathlib4/blob/e7cd8e7a8f5a7ba5473b9a89059741767319983d/Mathlib/Tactic/Block.lean#L29

The linked issue https://github.com/leanprover/lean4/issues/451 has been closed, so maybe that means `Mathlib/Tactic/Block` can now be updated somehow to avoid the need for these parens?

Reply:
See the lengthy discussion here: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.22.C2.B7.20tac.22.20block.20notation

AFAICT it's not fixed yet and Mario wants to use `∙` instead of `-` to resolve the issue.